util/test/execution_limiter.py

```diff
@@ -0,0 +1,107 @@
+#
+# The constructor takes the name of the executable and its timeout
+# value.
+class FileLock():
+    exec_name = ''
+    exec_timeout = ''
+    lock_file = ''
+
+    # simple backoff sleep, so that we have relatively fast latency for a bunch
+    # of quick running tests, but don't poll the machine continuously if a test
+    # is taking a while to run.
+    sleep_time = .1
+    def _backoff_sleep(self, max_sleep=.5, backoff_factor=1.1):
+	time.sleep(self.sleep_time)
+	if self.sleep_time < max_sleep:
+	    self.sleep_time *= backoff_factor
+   
+    def __init__(self, exec_name, exec_timeout):
+	lock_name = '{0}-chpl_program_executing'.format(getpass.getuser())
+        self.lock_file = os.path.join(tempfile.gettempdir(), lock_name)
+        self.exec_name = exec_name
+        self.exec_timeout = exec_timeout
+
+    def __enter__(self):
+        self._lock()
+
+    def __exit__(self, type, value, traceback):
+        self._unlock()
+
+    # parse the lock_file and return how long ago the lock_file was
+    # modified, the name of the test that has the lock, and the
+    # test's timeout value. If there's issues opening or parsing the
+    # file, return a default of: (1, "corrupted_lock_file", 0), so
+    # that the waiting test can continue.
+    def _parse_lock_file(self):
+        default_lock_file = (1, "corrupted_lock_file", 0)
+        try:
+            with open(self.lock_file, 'r') as f:
+                lines = f.read().splitlines()
+                other_filename = lines[0]
+                other_timeout = int(lines[1])
+
+                modified = time.time() - os.path.getmtime(self.lock_file)
+                return (modified, other_filename, other_timeout)
+        except:
+            log_info('Failed to parse lock file, using default one')
+            return default_lock_file
+
+    # Grab the execution "lock", spin waiting while some other test
+    # has the lock. If for some reason the other test didn't clean
+    # up the locking file, we'll continue if the lock hasn't been
+    # modified in the other test's timeout value.
+    def _lock(self):
+	modified = 0
+	other_timeout = 1
+	while os.path.exists(self.lock_file) and modified < other_timeout:
+            modified, other_filename, other_timeout = self._parse_lock_file()
+            log_info('"{0}" is waiting at most {1}s for "{2}" (started {3:.1f}s '
+                     'ago) to finish executing'.format(
+                     self.exec_name, other_timeout, other_filename, modified))
+            self._backoff_sleep()
+
+	# actually grab the "lock" by creating the file
+        log_info('Grabbing exec-lock for "{0}"'.format(self.exec_name))
+	with open(self.lock_file, 'w') as f:
+	    f.write('{0}\n{1}\n'.format(self.exec_name, self.exec_timeout))
+
+
+    def _unlock(self):
+	try:
+	    # release the "lock" by removing the file
+            log_info('Releasing exec-lock for "{0}"'.format(self.exec_name))
+	    os.remove(self.lock_file)
+	except:
+	    pass
```

Comment:
Makes sense. As noted in my comment below, stopping the data race would be nice, but not necessary for this PR.